src/qat/model/hardware_model.py

```diff
@@ -1,48 +1,124 @@
 from semver import Version
 
-from qat.model.autopopulate import AutoPopulate
-from qat.model.device import (
-    PhysicalBaseband,
-    PhysicalChannel,
-    PulseChannel,
-    Qubit,
-    Resonator,
-)
-from qat.model.serialisation import ComponentDict
+from qat.model.device import Qubit, QubitId
+from qat.utils.pydantic import WarnOnExtraFieldsModel
 
 VERSION = Version(0, 0, 1)
 
 
-class QuantumHardwareModel(AutoPopulate):
-    """
-    Base class for calibrating our QPU hardware.
+class LogicalHardwareModel(WarnOnExtraFieldsModel):
+    """Models a hardware with a given topology.
 
     Attributes:
-        version: The version of this hardware model.
-        physical_basebands: The physical basebands that model the LOs.
-        physical_channels: The physical channels that carry the pulses.
-        pulse_channels: The pulse channels on a particular device.
-        qubits: The superconducting qubits on the chip.
-        resonators: The resonators on the chip.
+        topology: Connectivity of the qubits in the hardware model.
     """
 
-    version: SemanticVersion = VERSION
-    physical_basebands: ComponentDict[PhysicalBaseband] = Field(frozen=True, default=dict())
-    physical_channels: ComponentDict[PhysicalChannel] = Field(frozen=True, default=dict())
-    pulse_channels: ComponentDict[PulseChannel] = Field(frozen=True, default=dict())
-    qubits: ComponentDict[Qubit] = Field(frozen=True, default=dict())
-    resonators: ComponentDict[Resonator] = Field(frozen=True, default=dict())
+    version: SemanticVersion = Field(frozen=True, repr=False, default=VERSION)
+    topology: dict[QubitId, set[QubitId]]
 
     @field_validator("version")
-    def version_compatibility(version):
+    def version_compatibility(version: Version):
         assert (
             version.major == VERSION.major
         ), f"Direct instantiation requires major version compatibility (expected {VERSION.major}.Y.Z, found {version})"
         assert (
             VERSION >= version
         ), f"Latest supported hardware model version {VERSION}, found {version}"
         return VERSION
+
+    def __eq__(self, other: LogicalHardwareModel) -> bool:
+        if type(self) != type(other):
+            return False
+
+        if self.model_fields != other.model_fields:
+            return False
+
+        if self.version != other.version:
+            return False
+
+        if self.topology != other.topology:
+            return False
+
+        return True
+
+    def __ne__(self, other: LogicalHardwareModel) -> bool:
+        return not self.__eq__(other)
+
+
+class QuantumHardwareModel(LogicalHardwareModel):
+    """Class for calibrating our QPU hardware.
+
+    Attributes:
+        qubits: The superconducting qubits on the chip.
+        constrained_topology: The connectivities between the qubits, which can be a subgraph of the physical topology.
+    """
+
+    qubits: dict[QubitId, Qubit]
```

Comment:
We _could_. However, I believe such solution would cause some overhead if we decide to remove some qubits from the hardware model due to them malfunctioning. In this case, there is no clear 1:1 mapping between the qubit index and the list index anymore..

Reply:
Ah thanks, that makes sense!